rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -1866,6 +1869,21 @@ impl Merger {
             // the batch at all.  Writing an empty batch currently panics
             if matched.num_rows() > 0 {
                 let row_ids = matched.column(row_id_col).as_primitive::<UInt64Type>();
+
+                let mut processed_row_ids = self.processed_row_ids.lock().unwrap();
+                for &row_id in row_ids.values() {
+                    if !processed_row_ids.insert(row_id) {
+                        return Err(datafusion::error::DataFusionError::Execution(
+                            format!(
+                                "Ambiguous merge insert: multiple source rows match the same target row with _rowid {}. \
+                                This could lead to data corruption. Please ensure each target row is matched by at most one source row.",
+                                row_id
+                            )
```

Comment:
I think the user would find it very valuable if we could provide the `on` column value here that matched multiple source rows. I *think* we should have enough context for that, right?